Add prop option to compact e/f func name

magodo opened the pull request:
Introduce a new property option: `compact_ef_func_name` to compact
expand/flatten function name.

E/F func name is constrcuted as: `resource_name` + `go_sdk_type_name`.
In some cases the `go_sdk_type_name` might use `resource_name` as
prefix, which in most case is redundent (cause without which, there is
no name conflict inside that go package). So as to make tf reviewer
happy, we can turn that option on.

This fixes issue: #60
<!-- 
Note: You may see "This branch is out-of-date with the base branch"
when you submit a pull request. This is fine! We don't use the GitHub
merge button to merge PRs, and you can safely ignore that message.

Thanks for contributing!
-->

magodo committed:
Add prop option to compact e/f func name

Introduce a new property option: `compact_ef_func_name` to compact
expand/flatten function name.

E/F func name is constrcuted as: `resource_name` + `go_sdk_type_name`.
In some cases the `go_sdk_type_name` might use `resource_name` as
prefix, which in most case is redundent (cause without which, there is
no name conflict inside that go package). So as to make tf reviewer
happy, we can turn that option on.

houkms commented:
Good improvement! I very much agree with this idea that we introduce a field to `PropertyOverride` to deal with expand/ flatten function name.

While I have a concern below about this PR

- It seems this PR only changes the ef funtion name for function definition. I could't find the codes changes about property field assigning. Can you help validate it?

DISCUSSION
- Even if users set the auto-compact function name, we can not deal with all kinds of situations that may appear (e.g. `expandArmWebApplicationFirewallPolicyWebApplicationFirewallCustomRule`).  How about we just leave this filed for users to flexibly customize the function name? (something like `custom_ef_func_name`) WDYT?

houkms reviewed (commented):
Good improvement! I very much agree with this idea that we introduce a field to `PropertyOverride` to deal with expand/ flatten function name.

While I have a concern below about this PR

- It seems this PR only changes the ef funtion name for function definition. I could't find the codes changes about property field assigning. Can you help validate it?

DISCUSSION
- Even if users set the auto-compact function name, we can not deal with all kinds of situations that may appear (e.g. `expandArmWebApplicationFirewallPolicyWebApplicationFirewallCustomRule`).  How about we just leave this filed for users to flexibly customize the function name? (something like `custom_ef_func_name`) WDYT?

magodo commented:
> It seems this PR only changes the ef funtion name for function definition. I could't find the codes changes about property field assigning. Can you help validate it?

This is my fault, i should have tested it (at least compile the generated code).

> Even if users set the auto-compact function name, we can not deal with all kinds of situations that may appear (e.g. expandArmWebApplicationFirewallPolicyWebApplicationFirewallCustomRule). How about we just leave this filed for users to flexibly customize the function name? (something like custom_ef_func_name) WDYT?

Yes, i agree. I assume that would be a mapping from the `go_sdk_type_name` to the final e/f func name, the `resource_name` can be omitted since the setting is already resided in a specific resource's definition/override.

magodo committed:
also modify func name on caller side

magodo committed:
cover more missing parts

magodo committed:
Provide custom ef name instead of compacting

Instead of compacting resource_name and go__type_name, allow user to
specify a custom ef func name.

houkms commented:
LGTM

houkms reviewed (commented):
LGTM
